README.md
# kurly

[![Build Status](https://gitlab.com/davidjpeacock/kurly/badges/master/build.svg)](https://gitlab.com/davidjpeacock/kurly/-/jobs)

kurly is an alternative to the widely popular curl program.

kurly is designed to operate in a similar manner to curl, with select features.
Notably, kurly is not aiming for feature parity, but common flags and mechanisms
particularly within the HTTP(S) realm are to be expected.

The current authors are not security experts, but want to contribute to the fledging
movement of replacing key tools and services with equivalents based on modern
and safe languages.  We recognize that people are fallible (including
ourselves), and for this reason believe we need all the help we can get.

Several languages exist which could be used to fulfill our goal, but in this case
we picked Golang.

## Installation

**Pre-requisite: Golang version 1.7.4 or higher.**

From source you can simply:

`go get gitlab.com/davidjpeacock/kurly`

## OS Distribution Package

`kurly` can be installed through package management systems on the following platforms:

* Arch Linux via Arch User Repos
  + For stable versions : `pacaur -S kurly` or `yaourt -S kurly`
  + For tip/development versions : `pacaur -S kurly-git` or `yaourt -S kurly-git` 

*If you're a package maintainer and you have prepared kurly for your OS of choice, please
PR this section.*

## Binary download

Binaries are provided for the following platforms:

* [Mac OS X amd64](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/darwin-amd64/kurly-darwin-amd64.tar.gz?job=compile)
* [Windows amd64](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/windows-amd64/kurly-windows-amd64.zip?job=compile)
* [Linux amd64](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/linux-amd64/kurly-linux-amd64.tar.gz?job=compile)
* [Linux arm](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/linux-arm/kurly-linux-arm.tar.gz?job=compile)
* [FreeBSD amd64](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/freebsd-amd64/kurly-freebsd-amd64.tar.gz?job=compile)
* [FreeBSD arm](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/freebsd-arm/kurly-freebsd-arm.tar.gz?job=compile)
* [OpenBSD amd64](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/openbsd-amd64/kurly-openbsd-amd64.tar.gz?job=compile)
* [OpenBSD arm](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/openbsd-arm/kurly-openbsd-arm.tar.gz?job=compile)
* [NetBSD amd64](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/netbsd-amd64/kurly-netbsd-amd64.tar.gz?job=compile)
* [NetBSD arm](https://gitlab.com/davidjpeacock/kurly/-/jobs/artifacts/master/raw/artifacts/netbsd-arm/kurly-netbsd-arm.tar.gz?job=compile)

## Usage

See `kurly --help` for usage information.

## Examples

Verbose output, showing headers
```
$ kurly -v https://httpbin.org/ip
*   Trying 23.23.171.5...
* TCP_NODELAY set
* Connected to httpbin.org (23.23.171.5) port 443 (#0)
* APLN, server accepted to use http/1.1
* TLSv1.2, TLS Handshake finished
* SSL connection using TLSv1.2 / ECDHE-RSA-AES-128-GCM-SHA256
* Server certificate:
*  subject: CN=httpbin.org
*  start date: Thu, 11 Jan 2018 23:37:29 UTC
*  expire date: Wed, 11 Apr 2018 23:37:29 UTC
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
> GET /ip HTTP/1.1
> User-Agent [Kurly/master]
> Accept [*/*]
> Host [httpbin.org]
< HTTP/1.1 200 OK
< Server [meinheld/0.6.1]
< Content-Type [application/json]
< Access-Control-Allow-Credentials [true]
< Content-Length [33]
< Via [1.1 vegur]
< Connection [keep-alive]
< Date [Mon, 12 Mar 2018 19:18:11 GMT]
< Access-Control-Allow-Origin [*]
< X-Powered-By [Flask]
< X-Processed-Time [0]
[<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<] 33 B/33 B
{
  "origin": "43.122.23.223"
}
```

Download file, preserving remote filename, timestamp, and following redirects
```
$ kurly -R -O -L http://cdimage.debian.org/debian-cd/current/amd64/iso-cd/debian-8.7.1-amd64-netinst.iso
[<<<<<<                                ] 41.2 MB/260 MB
```

Upload file
```
$ kurly -T ~/Downloads/image.jpeg https://httpbin.org/put
```

Posting elements with -d
```
$ kurly -d bingo=bongo https://httpbin.org/post
```

## Roadmap

Succinctly, we're planning to cover all curl-like features relevant to HTTP(S), and would
love you to help.

## Contributing

Bug reports, feature requests, and pull requests are all welcome.  Thank you!

Please see [CONTRIBUTING.md](https://gitlab.com/davidjpeacock/kurly/blob/master/CONTRIBUTING.md) for details of how to work with us.

## Maintainers

kurly is brought to you and maintained by:

* [Akilan Elango](https://gitlab.com/aki237)
* [Al S-M](https://github.com/alsm)
* [David J Peacock](https://gitlab.com/davidjpeacock)

## License

kurly is Copyright (c) 2017-2018 David J Peacock and Al S-M, and is published under the Apache 2.0 license.


CHANGELOG.md
# Changelog
This file contains a record of all non-trivial changes to kurly

## [Unreleased]

## [1.2.2] 20180704

## Added
* Gitlab CI/CD integration to build and archive the binary and associated files

## Fixed
* Uploading files and Sending request bodies
* Printing the headers of the Redirecting response
* Rebuilding the URL when it has no trailing slash
* Github references to Gitlab

## Removed
* Travis CI support
* snap package support
* rpm build spec file
* Dockerfile

## [1.2.1] 20180312

### Added
* Improved verbosity
* TLS Verbosity
* Support for insecure HTTPS
* Added man page
* Behind-the-scenes refactor for future maintenance
* Handle multiple URLs
* Snap installation via desktop UI

## [1.1.0] 20171229

### Added
* Resume transfer from offset
* Cookie and cookie jar support
* Send HTTP multipart post data
* Installation for Linux via snap package
* Installation for Arch Linux via Arch User Repos

## [1.0.0] 20170501

Initial public release


options.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"strings"
	"time"

	"github.com/alsm/ioprogress"
	"github.com/urfave/cli"
)

type Field struct {
	Type      string
	IsFile    bool
	Value     string
	Filealias string
}

type FormData map[string]Field

type Options struct {
	outputFilename string
	fileUpload     string
	remoteName     bool
	continueAt     string
	verbose        bool
	maxTime        uint
	remoteTime     bool
	cookie         string
	cookieJar      string
	followRedirect bool
	maxRedirects   uint
	redirectsTaken uint
	silent         bool
	method         string
	headers        []string
	agent          string
	user           string
	expectTimeout  uint
	data           []string
	dataAscii      []string
	dataRaw        []string
	dataBinary     []string
	dataURLEncode  []string
	form           []string
	head           bool
	insecure       bool
	fdata          FormData // fdata is the field for processed form data
}

func (o *Options) getOptions(app *cli.App) {
	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:        "output, o",
			Usage:       "Filename to name url content to",
			Destination: &o.outputFilename,
		},
		cli.StringFlag{
			Name:        "upload-file, T",
			Usage:       "File to upload",
			Destination: &o.fileUpload,
		},
		cli.BoolFlag{
			Name:        "remote-name, O",
			Usage:       "Save output to file named with file part of URL",
			Destination: &o.remoteName,
		},
		cli.StringFlag{
			Name:        "continue-at, C",
			Usage:       "Resume transfer from offset",
			Destination: &o.continueAt,
		},
		cli.BoolFlag{
			Name:        "v",
			Usage:       "Verbose output",
			Destination: &o.verbose,
		},
		cli.UintFlag{
			Name:        "max-time, m",
			Usage:       "Maximum time to wait for an operation to complete in seconds",
			Destination: &o.maxTime,
		},
		cli.BoolFlag{
			Name:        "R",
			Usage:       "Set the timestamp of the local file to that of the remote file, if available",
			Destination: &o.remoteTime,
		},
		cli.StringFlag{
			Name:        "cookie, b",
			Usage:       "Set the cookies to be sent along with this request",
			Destination: &o.cookie,
		},
		cli.StringFlag{
			Name:        "cookie-jar, c",
			Usage:       "File to which the cookies have to be written (in cURL's cookie-jar file format)",
			Destination: &o.cookieJar,
		},
		cli.BoolFlag{
			Name:        "location, L",
			Usage:       "Follow 3xx redirects",
			Destination: &o.followRedirect,
		},
		cli.UintFlag{
			Name:        "max-redirs",
			Usage:       "Maximum number of 3xx redirects to follow",
			Destination: &o.maxRedirects,
			Value:       10,
		},
		cli.BoolFlag{
			Name:        "silent, s",
			Usage:       "Mute kurly entirely, operation without any output",
			Destination: &o.silent,
		},
		cli.StringFlag{
			Name:        "request, X",
			Usage:       "HTTP method to use",
			Destination: &o.method,
			Value:       "GET",
		},
		cli.StringFlag{
			Name:        "user-agent, A",
			Usage:       "User agent to set for this request",
			Destination: &o.agent,
			Value:       "Kurly/" + version,
		},
		cli.StringFlag{
			Name:        "user, u",
			Usage:       "User authentication data to set for this request",
			Destination: &o.user,
		},
		cli.StringSliceFlag{
			Name:  "header, H",
			Usage: "Extra headers to be sent with the request",
		},
		cli.UintFlag{
			Name:        "expect100-timeout",
			Usage:       "Timeout in seconds for Expect: 100-continue wait period",
			Destination: &o.expectTimeout,
			Value:       1,
		},
		cli.StringSliceFlag{
			Name:  "data, d",
			Usage: "Sends the specified data in a POST request to the server",
		},
		cli.StringSliceFlag{
			Name:  "data-ascii",
			Usage: "The same as --data, -d",
		},
		cli.StringSliceFlag{
			Name:  "data-raw",
			Usage: "Basically the same as --data-binary (no @ interpretation)",
		},
		cli.StringSliceFlag{
			Name:  "data-binary",
			Usage: "Sends the data as binary",
		},
		cli.StringSliceFlag{
			Name:  "data-urlencode",
			Usage: "Sends the data as urlencoded ascii",
		},
		cli.StringSliceFlag{
			Name:  "form, F",
			Usage: "Send HTTP multipart post data",
		},
		cli.BoolFlag{
			Name:        "head, I",
			Usage:       "Get HEAD from URL only",
			Destination: &o.head,
		},
		cli.BoolFlag{
			Name:        "insecure, k",
			Usage:       "Allow insecure server connections when using TLS",
			Destination: &o.insecure,
		},
	}
}

func (o *Options) checkRedirect(req *http.Request, via []*http.Request) error {
	o.redirectsTaken++

	resp := req.Response

	if !o.followRedirect || o.redirectsTaken >= o.maxRedirects {
		return http.ErrUseLastResponse
	}

	if resp != nil {
		fmt.Fprintf(Incoming, "%s %s\n", resp.Proto, resp.Status)

		for k, v := range resp.Header {
			fmt.Fprintln(Incoming, k, v)
		}
		fmt.Fprintln(Incoming)
	}

	if o.verbose {
		Status.Println(" Ignoring the response body")
		Status.Printf(" Issuing request to this URL : %s\n", req.URL)
	}
	return nil
}

// BuildCommonOptions function is used to build all the common options from the cli.Context
// which is common for all the URL args passed from commandline.
// This this process takes place only once.
func (opts *Options) BuildCommonOptions(c *cli.Context) error {
	opts.headers = c.StringSlice("header")
	opts.user = c.String("user")
	opts.dataAscii = c.StringSlice("data")
	opts.dataAscii = append(opts.dataAscii, c.StringSlice("data-ascii")...)
	opts.dataBinary = c.StringSlice("data-binary")
	opts.dataRaw = c.StringSlice("data-raw")
	opts.dataURLEncode = c.StringSlice("data-urlencode")
	opts.form = c.StringSlice("form")

	// If verbose set the logs writers
	if opts.verbose {
		Incoming.(*LogWriter).SetOutput(os.Stderr)
		Outgoing.(*LogWriter).SetOutput(os.Stderr)
	}

	// Process form data or url-encoded data
	opts.ProcessData()
	d, err := opts.ProcessFormData()
	if err != nil {
		return err
	}
	if d != nil && len(opts.data) > 0 {
		return fmt.Errorf("only one type of body can be accepted : either multipart form or url encoded values")
	}
	opts.fdata = d

	// Set the request method if Head option is specified
	if opts.head {
		opts.method = "HEAD"
		Incoming = io.MultiWriter(os.Stdout, Incoming.(*LogWriter))
	}

	// Start the timer
	if opts.maxTime > 0 {
		maxTime(opts.maxTime)
	}

	return nil
}

// BuildTargetSpecificOptions function is used to build the options specific to a given URL target.
// This has to run for every URL separately.
func (opts *Options) BuildTargetSpecificOptions(target string) (io.Reader, error) {
	var body io.Reader
	// Set the output filename from the remote URL, if -O is passed.
	if opts.remoteName {
		opts.outputFilename = path.Base(target)
	}

	// Initialize the file upload if specified.
	if opts.fileUpload != "" {
		body = opts.uploadFile()
	}

	// Process headers and post data
	if len(opts.data) > 0 || len(opts.fdata) > 0 {
		var data bytes.Buffer
		opts.method = "POST"

		header := ""

		if len(opts.data) > 0 {
			header = "Content-Type: application/x-www-form-urlencoded"
			for i, d := range opts.data {
				data.WriteString(d)
				if i < len(opts.data)-1 {
					data.WriteRune('&')
				}
			}
		}

		if len(opts.fdata) > 0 {
			w := multipart.NewWriter(&data)
			for key, field := range opts.fdata {
				err := writeToMultipart(w, key, field)
				if err != nil {
					return nil, fmt.Errorf("unable to create http request; %s\n", err)
				}
			}
			w.Close()
			header = "Content-Type: " + w.FormDataContentType()
		}

		opts.headers = append([]string{header}, opts.headers...)
		body = &data
	}

	return body, nil
}

func (o *Options) ProcessData() {
	var uriEncodes url.Values
	for _, d := range o.dataAscii {
		parts := strings.SplitN(d, "=", 2)
		if len(parts) == 1 {
			if strings.HasPrefix(parts[0], "@") {
				data, err := ioutil.ReadFile(strings.TrimPrefix(parts[0], "@"))
				if err != nil {
					Status.Fatalf("Unable to read file %s : %s", strings.TrimPrefix(parts[0], "@"), err)
				}
				o.data = append(o.data, string(data))
				continue
			}
			o.data = append(o.data, d)
			continue
		}
		if strings.HasPrefix(parts[1], "@") {
			data, err := ioutil.ReadFile(strings.TrimPrefix(parts[1], "@"))
			if err != nil {
				Status.Fatalf("Unable to read file %s for data element %s\n", strings.TrimPrefix(parts[1], "@"), parts[0])
			}
			data = []byte(strings.Replace(string(data), "\r", "", -1))
			data = []byte(strings.Replace(string(data), "\n", "", -1))
			o.data = append(o.data, fmt.Sprintf("%s=%s", parts[0], string(data)))
		} else {
			o.data = append(o.data, d)
		}
	}
	for _, d := range o.dataRaw {
		parts := strings.SplitN(d, "=", 2)
		o.data = append(o.data, fmt.Sprintf("%s=%s", parts[0], parts[1]))
	}
	for _, d := range o.dataBinary {
		parts := strings.SplitN(d, "=", 2)
		o.data = append(o.data, fmt.Sprintf("%s=%s", parts[0], parts[1]))
	}
	for _, d := range o.dataURLEncode {
		parts := strings.SplitN(d, "=", 2)
		uriEncodes.Add(parts[0], parts[1])
	}
	if len(uriEncodes) > 0 {
		o.data = append(o.data, uriEncodes.Encode())
	}
}

func (o *Options) openOutputFile() *os.File {
	var err error
	var outputFile *os.File
	if o.outputFilename != "" {
		if outputFile, err = os.OpenFile(o.outputFilename, os.O_CREATE|os.O_RDWR, 0666); err != nil {
			Status.Fatalf("Error: Unable to create/open file '%s' for output\n", o.outputFilename)
		}
	} else {
		outputFile = os.Stdout
	}
	return outputFile
}

func (o *Options) uploadFile() io.Reader {
	o.method = "PUT"

	tr := &http.Transport{
		ExpectContinueTimeout: time.Duration(o.expectTimeout) * time.Second,
	}
	client.Transport = tr
	o.headers = append(o.headers, "Expect: 100-continue")

	reader, err := os.Open(o.fileUpload)
	if err != nil {
		Status.Fatalf("Error opening %s\n", o.fileUpload)
	}

	if !o.silent {
		fi, err := reader.Stat()
		if err != nil {
			Status.Fatalf("Unable to get file stats for %v\n", o.fileUpload)
		}
		return &ioprogress.Reader{
			Reader: reader,
			Size:   fi.Size(),
			DrawFunc: ioprogress.DrawTerminalf(os.Stderr, func(progress, total int64) string {
				return fmt.Sprintf(
					"%s %s",
					(ioprogress.DrawTextFormatBarWithIndicator(40, '>'))(progress, total),
					ioprogress.DrawTextFormatBytes(progress, total))
			}),
		}
	}

	return reader
}

// ProcessFormData is used to parse the form data passed as commandline arguments.
func (o *Options) ProcessFormData() (FormData, error) {
	if len(o.form) == 0 {
		return nil, nil
	}

	fd := make(FormData, 0)

	for _, field := range o.form {
		fieldName, f, err := parseField(field)
		if err != nil {
			return nil, err
		}

		fd[fieldName] = f
	}
	return fd, nil
}

func parseField(raw string) (string, Field, error) {
	f := Field{}
	returnKey := ""
	parts := splitFormParams(raw)
	for i, part := range parts {
		key, value, err := getKeyVal(part)
		if err != nil {
			return "", f, err
		}
		switch key {
		case "type":
			f.Type = value
		case "filename":
			f.Filealias = value
		default:
			if returnKey != "" || i != 0 {
				return "", Field{}, errors.New("malformed form data")
			}
			returnKey = key
			if strings.HasPrefix(value, "@") && len(value) > 1 {
				f.IsFile = true
				f.Value = value[1:]
				continue
			}

			f.Value = value
			f.IsFile = false
		}
	}
	if f.IsFile && f.Filealias == "" {
		f.Filealias = filepath.Base(f.Value)
	}

	return returnKey, f, nil
}

func splitFormParams(raw string) []string {
	inQuotes := false
	temp := ""
	strList := make([]string, 0)
	for _, ch := range raw {
		t := string(ch)
		if t == "\"" {
			inQuotes = !inQuotes
			continue
		}

		if t == ";" && !inQuotes {
			if temp != "" {
				strList = append(strList, temp)
				temp = ""
			}
			continue
		}

		temp += t
	}

	if temp != "" {
		strList = append(strList, temp)
	}

	return strList
}

func getKeyVal(raw string) (string, string, error) {
	splits := strings.Split(raw, "=")
	if len(splits) < 2 {
		return "", "", errors.New("not a valid key-value pair")
	}
	key := splits[0]
	val := strings.Join(splits[1:], "=")
	return key, val, nil
}


meta/build.sh
#!/bin/bash

# script taken from https://gist.github.com/mshafiee/5a681bbefda8f26f1f257d62f5e4a699



BIN_FILE_NAME_PREFIX=$1
PROJECT_DIR=$2
PLATFORMS="linux/amd64 linux/arm \
	   darwin/amd64 \
	   freebsd/amd64 freebsd/arm \
	   netbsd/amd64 netbsd/arm \
	   openbsd/amd64 openbsd/arm \
	   windows/amd64"

apt update
apt install -y zip

for PLATFORM in $PLATFORMS; do
        GOOS=${PLATFORM%/*}
        GOARCH=${PLATFORM#*/}
        FILEPATH="$PROJECT_DIR/artifacts/${GOOS}-${GOARCH}"
        #echo $FILEPATH
        mkdir -p $FILEPATH
        BIN_FILE_NAME="$FILEPATH/${BIN_FILE_NAME_PREFIX}"
        #echo $BIN_FILE_NAME
        if [[ "${GOOS}" == "windows" ]]; then BIN_FILE_NAME="${BIN_FILE_NAME}.exe"; fi
        CMD="CGO_ENABLED=0 GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-w -s' -o ${BIN_FILE_NAME}"
        #echo $CMD
        echo "${CMD}"
        eval $CMD || FAILURES="${FAILURES} ${PLATFORM}"
	if [[ "${GOOS}" == "windows" ]]; then
	    zip $FILEPATH/kurly-${GOOS}-${GOARCH}.zip -j ${BIN_FILE_NAME} LICENSE README.md
	    rm $FILEPATH/kurly.exe
	else
	    cp README.md LICENSE meta/kurly.man $FILEPATH
	    tar czvf $FILEPATH/kurly-${GOOS}-${GOARCH}.tar.gz -C $FILEPATH kurly README.md LICENSE kurly.man
	    rm $FILEPATH/kurly $FILEPATH/README.md $FILEPATH/LICENSE $FILEPATH/kurly.man
	fi
done


main.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/base64"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"mime/multipart"
	"net/http"
	"net/http/httptrace"
	"net/textproto"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/aki237/nscjar"
	"github.com/alsm/ioprogress"
	"github.com/urfave/cli"
)

var (
	client   = http.Client{}
	Status   = log.New(os.Stderr, "*", 0)
	Incoming io.Writer
	Outgoing io.Writer
)

const (
	version string = "1.2.2"
)

type LogWriter struct {
	*log.Logger
}

func (l *LogWriter) Write(p []byte) (n int, err error) {
	l.Print(string(p))
	return len(p), nil
}

func init() {
	cli.VersionFlag = cli.BoolFlag{
		Name:  "version, V",
		Usage: "print the version",
	}
	Incoming = &LogWriter{Logger: log.New(ioutil.Discard, "< ", 0)}
	Outgoing = &LogWriter{Logger: log.New(ioutil.Discard, "> ", 0)}
}

func main() {
	var opts Options

	app := cli.NewApp()
	app.Name = "kurly"
	app.Usage = "[options] URL"
	app.Version = version

	opts.getOptions(app)

	app.Action = func(c *cli.Context) error {
		if c.NArg() == 0 {
			cli.ShowAppHelp(c)
			os.Exit(0)
		}

		err := opts.BuildCommonOptions(c)
		if err != nil {
			return err
		}

		for _, uri := range c.Args() {
			err := fetchUrl(uri, opts, c)
			if err != nil {
				fmt.Fprintf(os.Stderr, "kurly : %s\n", err)
			}
		}
		return nil
	}

	app.Run(os.Args)
}

func fetchUrl(target string, opts Options, c *cli.Context) error {
	var remote *url.URL
	var err error
	var body io.Reader

	body, err = opts.BuildTargetSpecificOptions(target)
	if err != nil {
		return err
	}

	client.CheckRedirect = opts.checkRedirect
	if opts.insecure {
		client.Transport = &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		}
	}

	if remote, err = url.Parse(target); err != nil {
		return fmt.Errorf("Error: %s does not parse correctly as a URL", target)
	}

	if remote.Scheme == "" {
		remote.Scheme = "http"
		remote, _ = url.Parse(remote.String())
	}

	outputFile := opts.openOutputFile()

	if opts.method == http.MethodPut {
		// TODO : add support for reading contents from stdin
		if strings.HasSuffix(remote.Path, "/") || remote.Path == "" {
			remote.Path = filepath.Join(remote.Path, filepath.Base(opts.fileUpload))
			target = remote.String()
		}
	}

	continueAtInt := uint64(0)
	if opts.continueAt != "" {
		if opts.continueAt == "-" {
			fileInfo, err := outputFile.Stat()
			if err != nil {
				return fmt.Errorf("unable to set content range automatically from file; %s", err)
			}
			continueAtInt = uint64(fileInfo.Size())
		} else {
			continueAtInt, err = strconv.ParseUint(opts.continueAt, 10, 64)
			if err != nil {
				return fmt.Errorf("unable to create http request; expected a valid positive number for continue-at option")
			}
		}
	}

	req, err := http.NewRequest(opts.method, target, body)
	if err != nil {
		Status.Fatalf("Error: unable to create http %s request; %s\n", opts.method, err)
	}

	if opts.verbose {
		req = req.WithContext(httptrace.WithClientTrace(req.Context(), NewClientTraceForRequest(req)))
	}

	// Seek to given offset of the file and set the "Range" header
	if continueAtInt > 0 {
		if opts.outputFilename != "" {
			_, err = outputFile.Seek(int64(continueAtInt), 0)
			if err != nil {
				Status.Fatalf("Error: seek in the given output file; %s\n", err)
			}
		}
		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", continueAtInt))
	}
	req.Header.Set("User-Agent", opts.agent)
	if opts.user != "" {
		req.Header.Set("Authorization", "Basic "+encodeToBase64(opts.user))
	}
	req.Header.Set("Accept", "*/*")
	req.Header.Set("Host", remote.Host)
	if body != nil {
		switch b := body.(type) {
		case *os.File:
			fi, err := b.Stat()
			if err != nil {
				Status.Fatalf("Unable to get file stats for %v\n", opts.fileUpload)
			}
			req.ContentLength = fi.Size()
			req.Header.Set("Content-Length", strconv.FormatInt(fi.Size(), 10))
		case *ioprogress.Reader:
			req.ContentLength = b.Size
			req.Header.Set("Content-Length", strconv.FormatInt(b.Size, 10))
		case *bytes.Buffer:
			req.Header.Set("Content-Length", strconv.FormatInt(int64(b.Len()), 10))
		}
	}
	setHeaders(req, opts.headers)
	setCookieHeader(req, opts.cookie)

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if continueAtInt > 0 && resp.StatusCode == 416 {
		return fmt.Errorf("unable to get URL; %s\n", "Either the server doesn't support ranges or an invalid range is passed")
	}

	fmt.Fprintf(Incoming, "%s %s\n", resp.Proto, resp.Status)

	for k, v := range resp.Header {
		fmt.Fprintln(Incoming, k, v)
	}

	fmt.Fprintln(Incoming)

	if !opts.head {
		if !opts.silent {
			progressR := &ioprogress.Reader{
				Reader: resp.Body,
				Size:   resp.ContentLength,
				DrawFunc: ioprogress.DrawTerminalf(os.Stderr, func(progress, total int64) string {
					return fmt.Sprintf(
						"%s %s",
						(ioprogress.DrawTextFormatBarWithIndicator(40, '<'))(progress, total),
						ioprogress.DrawTextFormatBytes(progress, total))
				}),
			}
			if _, err = io.Copy(outputFile, progressR); err != nil {
				return fmt.Errorf("failed to copy URL content; %s", err)
			}
		}
		if opts.silent {
			if _, err = io.Copy(outputFile, resp.Body); err != nil {
				return fmt.Errorf("failed to copy URL content; %s", err)
			}
		}
	}

	if opts.outputFilename != "" {
		outputFile.Close()
	}

	if rTime := resp.Header.Get("Last-Modified"); opts.remoteTime && rTime != "" {
		if t, err := time.Parse("Mon, 02 Jan 2006 15:04:05 MST", rTime); err == nil {
			os.Chtimes(opts.outputFilename, t, t)
		}
	}

	if opts.cookieJar != "" && len(resp.Cookies()) > 0 {
		cookies := resp.Cookies()
		for _, val := range cookies {
			if val.Domain == "" {
				u, err := resp.Location()
				if err != nil {
					u = req.URL
				}

				val.Domain = u.Hostname()
			}
		}
		saveCookies(cookies, opts.cookieJar)
	}
	return nil
}

func setHeaders(r *http.Request, h []string) {
	for _, header := range h {
		hParts := strings.SplitN(header, ":", 2)
		switch len(hParts) {
		case 0:
			//surely not
		case 1:
			//must be an empty Header or a delete
			switch {
			case strings.HasSuffix(header, ";"):
				r.Header.Set(strings.TrimSuffix(header, ";"), "")
			case strings.HasSuffix(header, ":"):
				r.Header.Del(strings.TrimSuffix(header, ":"))
			default:
			}
		case 2:
			//standard expected
			r.Header.Set(strings.TrimSpace(hParts[0]), strings.TrimSpace(hParts[1]))
		default:
			// No possible.
		}
	}
}

func writeToMultipart(w *multipart.Writer, key string, field Field) error {
	if field.IsFile {
		file, err := os.Open(field.Value)
		if err != nil {
			return err
		}
		defer file.Close()

		h := make(textproto.MIMEHeader)
		h.Set("Content-Disposition",
			fmt.Sprintf(`form-data; name="%s"; filename="%s"`,
				key, field.Filealias))
		h.Set("Content-Type", "application/octet-stream")
		if field.Type != "" {
			h.Set("Content-Type", field.Type)
		}

		fw, err := w.CreatePart(h)
		if err != nil {
			return err
		}
		if _, err = io.Copy(fw, file); err != nil {
			return err
		}
	} else {
		h := make(textproto.MIMEHeader)
		h.Set("Content-Disposition",
			fmt.Sprintf(`form-data; name="%s"`, key))
		if field.Type != "" {
			h.Set("Content-Type", field.Type)
		}
		wr, err := w.CreatePart(h)
		if err != nil {
			return err
		}
		_, err = wr.Write([]byte(field.Value))
		if err != nil {
			return err
		}

	}
	return nil
}

func setCookieHeader(r *http.Request, arg string) {
	// according to cURL man pages and operations, if the cookie string passed has
	// a "=" in it, it means that is a valid cookie. Else it will search for the
	// cookie jar file of the name and tries to read cookies from it
	if strings.Contains(arg, "=") {
		r.Header.Set("Cookie", arg)
		return
	}

	// If the specified cookie string is invalid and the the file of the same name
	// doesn't exist, cURL doesn't throw any error, rather it just sets the cookie
	// to null
	cookieFile, err := os.Open(arg)
	if err != nil {
		return
	}

	p := nscjar.Parser{}

	cookies, err := p.Unmarshal(cookieFile)
	if err != nil {
		return
	}

	// reference time for checking the expiry of a cookie from the file
	t := time.Now()

	for _, val := range cookies {
		// Skip if cookie is secure and request isn't
		if val.Secure && r.URL.Scheme != "https" {
			continue
		}

		// Skip if the cookie has expired
		if val.Expires.Unix() <= t.Unix() {
			continue
		}

		host := r.URL.Hostname()
		// Domain matching (according to rfc6265::Section-5.1.3)
		if strings.ToLower(val.Domain) != strings.ToLower(host) {
			suffix := val.Domain
			if !strings.HasPrefix(val.Domain, ".") {
				suffix = "." + suffix
			}
			if !strings.HasSuffix(host, suffix) {
				continue
			}
		}

		// Path matching (according to rfc6265::Section-5.1.4)
		uri, _ := url.Parse("http://" + val.Domain + val.Path)
		if strings.HasPrefix(r.URL.Path, uri.Path) && r.URL.Path != uri.Path {
			continue
		}

		r.AddCookie(val)
	}
}

func saveCookies(cs []*http.Cookie, filename string) {
	f, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR, 0666)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
		return
	}

	jar := nscjar.NewCookieJar()

	// get the already stored cookies if exists in file.
	p := nscjar.Parser{}
	if cookies, err := p.Unmarshal(f); err == nil {
		jar.AddCookies(cookies...)
	}

	jar.AddCookies(cs...)

	_, err = f.Seek(0, 0)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
		return
	}

	err = jar.Marshal(f)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
	}
	err = f.Close()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
	}
}

func maxTime(maxTime uint) {
	go func() {
		<-time.After(time.Duration(maxTime) * time.Second)
		Status.Fatalf("Error: Maximum operation time of %d seconds expired, aborting\n", maxTime)
	}()
}

func encodeToBase64(a string) string {
	return base64.StdEncoding.EncodeToString([]byte(a))
}
